File: 2023/D01.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from pathlib import Path
import re

## Current file direction
cfd = Path(__file__).parent.absolute()

## Read input
f = open(str(cfd)+'/data/D01_input.txt', 'r')
lines = f.readlines()

## Part one counts
cval = 0 # sum of calibration values
for l in lines:
    # print(l.split()[0])
    digits = [x for x in l.split()[0] if x.isdigit()]
    ## Sum up
    cval += int(digits[0] + digits[-1])

## Part one answer
print("The sum of all of the calibration values is {}.\n".format(cval))

## Part two preparation
num = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]
dict_num = {n: i+1 for i, n in enumerate(num)}
for i in range(9):
    dict_num[str(i+1)] = i+1
# print(dict_num)
num_all = list(dict_num.keys())
pattern = num_all[0]
for n in num_all:
    pattern += "|"+n
# print(pattern)

## Part two pre-test
example = "7twoneight8nnine" # expected: 721889
print("Example: ", example)
print("Expected output: 721889\nActual output: ", 
      "".join([re.sub(x, str(dict_num[x]), x) for x in re.findall(rf'(?=({pattern}))', example)]), 
      "\n")

## Part two counts
cval2 = 0 # corrected sum of calibration values
for l in lines:
    ## Digitize each line
    digits = [re.sub(x, str(dict_num[x]), x) for x in re.findall(rf'(?=({pattern}))', l)]
    # print(l, digits, '\n')
    ## Sum up
    cval2 += int(digits[0] + digits[-1])

## Part two answer
print("The corrected sum of all of the calibration values is {}.\n".format(cval2))

File: 2023/D01.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.Arrays;

public class D01 {

  public static void main(String[] args) {
    String inputFile = "data/D01_input.txt";
    int sum = 0;
    int sumCorrected = 0;

    try (BufferedReader br = new BufferedReader(new FileReader(inputFile))) {
      String line;
      while ((line = br.readLine()) != null) {
        // Part one counts
        int firstDigit = extractFirstDigit(line);
        int lastDigit = extractLastDigit(line);
        int calibrationValue = combineDigits(firstDigit, lastDigit);
        sum += calibrationValue;

        // Part two counts
        String lineCorrected = convertToNumeric(line); // all to digits
        int firstDigitCorrected = extractFirstDigit(lineCorrected);
        int lastDigitCorrected = extractLastDigit(lineCorrected);
        int calibrationValueCorrected = combineDigits(firstDigitCorrected, lastDigitCorrected);
        sumCorrected += calibrationValueCorrected;
      }
    } catch (IOException e) {
      e.printStackTrace();
    }

    // Part one result
    System.out.println("Sum of all the calibration values: " + sum);
    System.out.println();

    // Part two pre-test
    String example = "7twoneight8nnine";
    System.out.println("Example: " + example);
    System.out.println("Expected output: " + "721889");
    String convertedLine = convertToNumeric(example);
    System.out.println("Actual output: " + convertedLine);
    System.out.println();

    // Part two result
    System.out.println("Sum of all the calibration values (lettre-digits corrected): " + sumCorrected);
    System.out.println();
  }

  private static String convertToNumeric(String line) {
    // Define mappings for spelled-out numbers to digits
    String[] numbers = {"one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};

    // Replace spelled-out numbers with digits, favoring the first appearance
    boolean foundDigit = true; // boost the loop
    while (foundDigit) {
      foundDigit = false; //pause the loop

      int firstIndexInLine = line.length(); // initialize to the length of the line
      String numberToReplace = null;
      // Check each spelled-out number
      for (String number : numbers) {
        // Check if the current spelled-out number exists in the line
        if (line.contains(number)) {
          // Find the index of the current spelled-out number in the line
          int index = line.indexOf(number.charAt(0));
          // If the number is found and its index is smaller than the smallest index found so far
          if (index != -1 && index < firstIndexInLine) {
            firstIndexInLine = index;
            numberToReplace = number;
            foundDigit = true; // set foundDigit to true to continue the loop
          }
        }
      }
      // If a spelled-out number was found, replace it with its corresponding digit
      if (foundDigit) {
        String digitString = String.valueOf(Arrays.asList(numbers).indexOf(numberToReplace) + 1);
        line = line.replaceFirst(numberToReplace, digitString);
        // System.out.println(line);
      }
    }

    return line;
  }

  private static int extractFirstDigit(String line) {
    for (char c : line.toCharArray()) {
      if (Character.isDigit(c)) {
        return Character.getNumericValue(c);
      }
    }
    return -1; // no digit found
  }

  private static int extractLastDigit(String line) {
    for (int i = line.length() - 1; i >= 0; i--) {
      char c = line.charAt(i);
      if (Character.isDigit(c)) {
        return Character.getNumericValue(c);
      }
    }
    return -1; // no digit found
  }

  private static int combineDigits(int firstDigit, int lastDigit) {
    if (firstDigit == -1 || lastDigit == -1) {
      return 0; // return 0 if any of the digits is not found
    }
    return firstDigit * 10 + lastDigit;
  }

}
